AsyncDisplayKit/ASDisplayNode+Yoga.mm

```diff
@@ -0,0 +1,377 @@
+  }
+}
+
+YGAlign yogaAlignSelf(ASStackLayoutAlignSelf alignSelf)
+{
+  switch (alignSelf) {
+    case ASStackLayoutAlignSelfStart:   return YGAlignFlexStart;
+    case ASStackLayoutAlignSelfCenter:  return YGAlignCenter;
+    case ASStackLayoutAlignSelfEnd:     return YGAlignFlexEnd;
+    case ASStackLayoutAlignSelfStretch: return YGAlignStretch;
+    case ASStackLayoutAlignSelfAuto:    return YGAlignAuto;
+  }
+}
+
+YGFlexDirection yogaFlexDirection(ASStackLayoutDirection direction)
+{
+  return direction == ASStackLayoutDirectionVertical ? YGFlexDirectionColumn : YGFlexDirectionRow;
+}
+
+float yogaFloatForCGFloat(CGFloat value)
+{
+  if (value < CGFLOAT_MAX / 2) {
+    return value;
+  } else {
+    return YGUndefined;
+  }
+}
+
+float yogaDimensionToPoints(ASDimension dimension)
+{
+  ASDisplayNodeCAssert(dimension.unit == ASDimensionUnitPoints,
+                       @"Dimensions should not be type Fraction for this method: %f", dimension.value);
+  return yogaFloatForCGFloat(dimension.value);
+}
+
+float yogaDimensionToPercent(ASDimension dimension)
+{
+  ASDisplayNodeCAssert(dimension.unit == ASDimensionUnitFraction,
+                       @"Dimensions should not be type Points for this method: %f", dimension.value);
+  return 100.0 * yogaFloatForCGFloat(dimension.value);
+
+}
+
+ASDimension dimensionForEdgeWithEdgeInsets(YGEdge edge, ASEdgeInsets insets)
+{
+  switch (edge) {
+    case YGEdgeLeft:   return insets.left;
+    case YGEdgeTop:    return insets.top;
+    case YGEdgeRight:  return insets.right;
+    case YGEdgeBottom: return insets.bottom;
+    default: ASDisplayNodeCAssert(NO, @"YGEdge other than ASEdgeInsets is not supported.");
+      return ASDimensionAuto;
+  }
+}
+
+YGSize ASLayoutElementYogaMeasureFunc(YGNodeRef yogaNode, float width, YGMeasureMode widthMode,
+                                      float height, YGMeasureMode heightMode)
+{
+  id <ASLayoutElement> layoutElement = (__bridge id <ASLayoutElement>)YGNodeGetContext(yogaNode);
+  ASSizeRange sizeRange;
+  sizeRange.max = CGSizeMake(width, height);
+  sizeRange.min = sizeRange.max;
+  if (widthMode == YGMeasureModeAtMost) {
+    sizeRange.min.width = 0.0;
+  }
+  if (heightMode == YGMeasureModeAtMost) {
+    sizeRange.min.height = 0.0;
+  }
+  CGSize size = [[layoutElement layoutThatFits:sizeRange] size];
+  return (YGSize){ .width = (float)size.width, .height = (float)size.height };
+}
+
+#pragma mark - ASDisplayNode+Yoga
+
+@interface ASDisplayNode (YogaInternal)
+@property (nonatomic, weak) ASDisplayNode *yogaParent;
+@property (nonatomic, assign) YGNodeRef yogaNode;
+@end
+
+@implementation ASDisplayNode (Yoga)
+
+- (void)setYogaParent:(ASDisplayNode *)yogaParent
+{
+  _yogaParent = yogaParent;
+  if (yogaParent) {
+    self.hierarchyState |= ASHierarchyStateYogaLayoutEnabled;
+  } else {
+    self.hierarchyState &= ~ASHierarchyStateYogaLayoutEnabled;
+  }
+}
+
+- (ASDisplayNode *)yogaParent
+{
+  return _yogaParent;
```

Comment:
I realize these methods don't currently lock.  This is something I have in mind and will follow up on.  For now, there is a risk of deadlocking if I add locks, and so I would like to approach this with a careful analysis of the correct locking before dropping them in.